micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastAdapter.java

```diff
@@ -0,0 +1,196 @@
+package io.micrometer.core.instrument.binder.cache;
+
+class HazelcastAdapter {
+    static String nameOf(Object cache) {
+        try {
+            return (String) iMapClass().getMethod("getName").invoke(cache);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Class iMapClass() throws ClassNotFoundException {
+        try {
+            return Class.forName("com.hazelcast.map.IMap");
+        } catch (ClassNotFoundException e) {
+            return Class.forName("com.hazelcast.core.IMap");
+        }
+    }
+
+    private static Class localMapStatsClass() throws ClassNotFoundException {
+        try {
+            return Class.forName("com.hazelcast.map.LocalMapStats");
+        } catch (ClassNotFoundException e) {
+            return Class.forName("com.hazelcast.monitor.LocalMapStats");
+        }
+    }
+
+    private static Class nearCacheStatsClass() throws ClassNotFoundException {
+        try {
+            return Class.forName("com.hazelcast.nearcache.NearCacheStats");
+        } catch (ClassNotFoundException e) {
+            return Class.forName("com.hazelcast.monitor.NearCacheStats");
+        }
+    }
+
+    static class IMap<K, V> {
```

Comment:
This implementation could just implement the original IMap

Reply:
if it implements the original IMap, then doesn't it need the IMap class at compile time?

Reply:
🤦‍♀️

Reply:
For a second I thought we could avoid dropping `C extends IMap<K, V>` generic bounds, but yeah - that's a no-go